interface/src/avatar/MyAvatar.cpp

```diff
@@ -1564,6 +1583,33 @@ void MyAvatar::sanitizeAvatarEntityProperties(EntityItemProperties& properties)
     properties.markAllChanged();
 }
 
+void MyAvatar::addAvatarEntitiesToTree() {
+    AvatarEntityMap::const_iterator constItr = _cachedAvatarEntityBlobs.begin();
+    while (constItr != _cachedAvatarEntityBlobs.end()) {
+        QUuid id = constItr.key();
+        _entitiesToAdd.push_back(id);  // worked once: hat shown. then unshown when permissions removed but then entity was deleted somewhere along the line!
+        ++constItr;
+    }
+}
+
+bool MyAvatar::hasAvatarEntities() const {
+    return _cachedAvatarEntityBlobs.count() > 0;
+}
+
+void MyAvatar::handleCanRezAvatarEntitiesChanged(bool canRezAvatarEntities) {
+    if (canRezAvatarEntities) {
+        // Start displaying avatar entities.
+        // Allow time for the avatar mixer to be updated with the user's permissions so that it doesn't discard the avatar 
+        // entities sent.
+        QTimer::singleShot(2 * DOMAIN_SERVER_CHECK_IN_MSECS, this, [this] {
```

Comment:
couldn’t it take longer?  this seems hacky

Reply:
Here's what happens:
- Domain owner enables canRezAvatarEntities permission for a user.
- User's Interface sends a `DomainListRequest` protocol message to the domain server every `DOMAIN_SERVER_CHECK_IN_MSECS` as a "domain server check-in".
- Avatar mixer assignment client sends a `DomainListRequest` protocol message to the domain server every `DOMAIN_SERVER_CHECK_IN_MSECS` as a "domain server check-in".
- In response to each `DomainListRequest`, the domain server sends a `DomainList` protocol message. This message contains the latest permissions settings.
- User's Interface processes `DomainList` protocol message and sees that avatar entities are now enabled for user, so adds user's avatar entities to tree which causes them to be sent to the avatar mixer.
- Avatar mixer processes `DomainList` protocol message and updates its copy of the domain permissions including the updated permission for the user.
- Avatar mixer receives avatar entities from user. If the avatar entities arrive before the `DomainList` protocol message with the user's udated permissions, the avatar entities are discarded and so not displayed to other users.

In the domain server the processing `DomainListRequest` messages there could be up to `DOMAIN_SERVER_CHECK_IN_MSECS` delay between the domain server sending updated permissions to the user and sending updated permissions to the avatar mixer, assuming that incoming `DomainListRequest` messages are being received regularly. 
In practice, the domain server and avatar mixer will have a fast connection between them (same machine or local network) so they shudln't be subject to networking delays.
There may be longer delays and variability in the connection between the user's Interface and the domain server. However, the worst case scenario is that there is no delay (e.g., Interface is on the same machine as the domain server) and Interface's `DomainListRequest` request comes in at `DOMAIN_SERVER_CHECK_IN_MSECS` before the avatar mixer's `DomainListRequest` request, so Inteface potentially knows that it has canRezAvatarEntities permissions `DOMAIN_SERVER_CHECK_IN_MSECS` before the avatar mixer does.

So, we could in theory get away with having just a `1 * DOMAIN_SERVER_CHECK_IN_MSECS` delay before sending the avatar entities to the avatar mixer, but a longer delay doesn't hurt.

Reply:
but none of those times are really guaranteed, right?  a slow client could be sending or processing check ins slower, and a slow mixer could be responding slowly, especially if there are lots of clients.  is there a way we can detect “we got the next check in, we’re good to go now” and move this code to the avatar update or something like that?

Reply:
If the Interface client is slow: This doesn't matter: Interface can send its avatar entities late; they'll just display late for observers.

If the avatar mixer is slow: It it's bogged down it's still likely that incoming avatar entities from Interface will be processed after the incoming `DomainList` message from the domain server that enabled avatar entities for the user. If the avatar mixer is bogged down so much that it can't send `DomainListRequest` messages to the domain server in a timely manner nor process the `DomainList` response in a timely manner, then the user's avatar entities won't display for observers unless the user teleports away and back or relogs.

In order to address the avatar-mixer-bogged-down-bigly scenario, Interface needs to hold off sending its avatar entities until it knows that the avatar mixer has received a `DomainList` message sent after the `DomainList` that interface received granting it avatar entities permissions. This could be achieved using the timestamp already included in the `DomainList` message, and making the avatar mixer send the timestamp value to Interface. I haven't found any suitable protocol messages to add the timestamp value to (protocol change); it would probably be better to add a new message (protocol change).

However, checking the timestamps in this manner seems like overkill for this scenario? Perhaps a simpler but more conservative solution could be to base Interface's send delay on, say, the greater of `5 * DOMAIN_SERVER_CHECK_IN_MSECS` and `3 * avatarPing` (3 * the avatar mixer ping time)? 
Note: Should also cancel the send delay timer if the user's permissions revert to "can't rez avatar entities" during the delay.

Reply:
ok, I'm convinced, I think maybe just a comment summarizing some of this would be sufficient and we can revisit in the future if bugs crop up, but the value probably won't be changing often anyways